src/tasks/LargeFileUploadTask.ts

```diff
@@ -0,0 +1,365 @@
+   * @param {ErrorMappings} [errorMappings] - error mappings.
+   *
+   * @throws {Error} If any of the required parameters are undefined or invalid.
+   */
+  constructor(
+    readonly requestAdapter: RequestAdapter,
+    readonly uploadSession: Parsable,
+    readonly uploadStreamProvider: () => ReadableStream<Uint8Array>,
+    readonly maxSliceSize = -1,
+    readonly parsableFactory: ParsableFactory<T>,
+    errorMappings: ErrorMappings,
+  ) {
+    if (!uploadSession) {
+      const error = new Error("Upload session is undefined, Please provide a valid upload session");
+      error.name = "Invalid Upload Session Error";
+      throw error;
+    }
+    if (!uploadStreamProvider) {
+      const error = new Error("Upload stream provider is undefined, Please provide a valid upload stream");
+      error.name = "Invalid Upload Stream Error";
+      throw error;
+    }
+    if (!requestAdapter) {
+      const error = new Error("Request adapter is undefined, Please provide a valid request adapter");
+      error.name = "Invalid Request Adapter Error";
+      throw error;
+    }
+    if (!parsableFactory) {
+      const error = new Error("Parsable factory is undefined, Please provide a valid parsable factory");
+      error.name = "Invalid Parsable Factory Error";
+      throw error;
+    }
+    if (maxSliceSize <= 0) {
+      this.maxSliceSize = DefaultSliceSize;
+    }
+    this.parsableFactory = parsableFactory;
+
+    this.Session = this.extractSessionInfo(uploadSession);
+    this.rangesRemaining = this.getRangesRemaining(this.Session);
+    this.errorMappings = errorMappings;
+  }
+
+  /**
+   * Uploads the file in a sequential order by slicing the file in terms of ranges.
+   *
+   * @param {IProgress} [progress] - Optional progress receiver to report upload progress.
+   * @returns {Promise<UploadResult<T>>} - The result of the upload.
+   * @throws {Error} If the upload fails.
+   */
+  public async upload(progress?: IProgress): Promise<UploadResult<T>> {
+    const sliceRequests = this.getUploadSliceRequests();
+    for (const request of sliceRequests) {
+      try {
+        const uploadResult = await this.uploadWithRetry(request);
+        progress?.report(request.rangeEnd);
+        const { itemResponse, location } = uploadResult as Partial<UploadResult<T>>;
+        if (itemResponse || location) {
+          return uploadResult as UploadResult<T>;
+        }
+      } catch (e) {
+        console.error(e);
+      }
+    }
+    throw new Error("Upload failed");
+  }
+
+  /**
+   * Uploads a slice with retry logic.
+   *
+   * @param {UploadSlice<T>} uploadSlice - The upload slice to be uploaded.
+   * @param {number} [maxTries=3] - The maximum number of retry attempts.
+   * @returns {Promise<UploadResult<T> | UploadSession | undefined>} - The result of the upload.
+   * @throws {Error} If the maximum number of retries is reached.
+   */
+  private async uploadWithRetry(
```

Comment:
I think we already had a comment for that by retry should be a concern of the middleware handler and we shouldn't mix concerns here.